hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -682,11 +681,27 @@ async def listfiles(self, url: str, recursive: bool = False) -> AsyncIterator[Fi
         except StopAsyncIteration:
             raise FileNotFoundError(url)  # pylint: disable=raise-missing-from
 
+        async def is_file(entry):
+            if await entry.is_dir():
+                return False
```

Comment:
So I think `is_file` is a bad name.  This should be something like `skip` or `should_yield`?  Because you want to return directories here.  You only want to skip empty-file-directories.

Reply:
Why would I want to return directories? We're listing files (or at least that's what I thought it was). 

I also broke listfiles in other tests:

```
        async def listfiles(dir, recursive):
            return {(await entry.url(), await entry.is_file()) async for entry in await fs.listfiles(dir, recursive)}
    
        assert await listfiles(f'{base}foo/', recursive=True) == {(a, True), (c, True)}
>       assert await listfiles(f'{base}foo/', recursive=False) == {(a, True), (b, False)}
E       AssertionError: assert {('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/a', True)} == {('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/a', True),\n ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/b/', False)}
E         Extra items in the right set:
E         ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/b/', False)
E         Full diff:
E           {
E            ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/a',
E             True),
E         -  ('gs://hail-test-dmk9z/tmp/QPTOQ0djdUWM3WYNHTbKew/foo/b/',
E         -   False),
E           }
```

```
        assert await listfiles(f'{base}foo/', recursive=True) == {(a, True), (c, True)}
>       assert await listfiles(f'{base}foo/', recursive=False) == {(a, True), (b, False)}
E       AssertionError: assert {('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/a', True)} == {('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/a', True),\n ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/b/', False)}
E         Extra items in the right set:
E         ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/b/', False)
E         Full diff:
E           {
E            ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/a',
E             True),
E         -  ('gs://hail-test-dmk9z/tmp/fOxVwBRMRHj37hvn8H20i1/foo/b/',
E         -   False),
E           }

```

Reply:
Maybe these tests need to be fixed rather than my code being wrong.

Reply:
Hmm, no, the code is definitely wrong.  `listfiles(..., recursive=False)` should definitely be returning **directories**.

Reply:
Let me clarify a bit:

`listfiles(..., recursive=True)` automatically descends into directories and only returns files.

`listfiles(..., recursive=False)` acts like listing directories on UNIX, and only returns things directly inside a directory, whether it be files or more directories.

Google Storage listing API supports both of these modes directly.  UNIX directory listing only implements `recursive=False`, and we iterate recursively through directories explicit in the LocalAsyncFS backend.

We don't make much use of `recursive=False`, but I think users will want it when we expose this to them (replacing hl.hadoop_ls, etc.)

Does that help?

Reply:
Obviously, I need to document the AsyncFS API ASAP.

Reply:
And maybe rename `listfiles` to `listdir`?  We're listing the contents of a directory.

Reply:
I think the equivalent operations of what we're doing is listfiles_flat is the equivalent of scandir in Python. listfiles_recursive doesn't have an equivalent operation in unix or python. I looked at `ls -R` but that lists both file and directory entries. `listfiles` makes sense to me for `listfiles_recursive`. `listdir` makes sense to me for `listfiles_flat`. Can we have both operations be top-level things rather than the `recursive=True` flag?

Reply:
> Can we have both operations be top-level things rather than the recursive=True flag?

Yeah, definitely.  I'll put it on the list of things to cleanup and refine when I start documenting.